modular_skyrat/modules/reagent_forging/code/forge.dm

```diff
@@ -133,556 +173,523 @@
+	used_tray = tray
+	in_use = TRUE // You can't use the forge if there's a tray sitting in it
+	update_appearance()
 
-		for(var/datum/reagent/weapon_reagent in attacking_weapon.reagents.reagent_list)
-			if(weapon_reagent.volume < MIN_IMBUE_REQUIRED)
-				attacking_weapon.reagents.remove_all_type(weapon_reagent.type)
-				continue
+/// Take the used_tray and spit it out, updating everything relevant to that
+/obj/structure/reagent_forge/proc/remove_tray_from_forge(mob/living/carbon/user)
+	if(!used_tray)
+		if(user)
+			balloon_alert_to_viewers("no tray")
+		return
 
-			if(is_type_in_typecache(weapon_reagent, disallowed_reagents))
-				fail_message(user, "The enchanted flames of the forge rebuke your attempt to work [weapon_reagent.name] into [attacking_weapon]...")
-				attacking_weapon.reagents.remove_all_type(weapon_reagent.type)
-				continue
+	if(user)
+		user.put_in_hands(used_tray)
+		balloon_alert_to_viewers("removed [used_tray]")
+	else
+		used_tray.forceMove(get_turf(src))
+	used_tray = null
+	in_use = FALSE
 
-			weapon_component.imbued_reagent += weapon_reagent.type
-			attacking_weapon.name = "[weapon_reagent.name] [attacking_weapon.name]"
+/// Adds to either the strong or weak fuel timers from the given stack
+/obj/structure/reagent_forge/proc/refuel(obj/item/stack/refueling_stack, mob/living/user, is_strong_fuel = FALSE)
+	in_use = TRUE
 
-		if(attacking_weapon.name == initial(attacking_weapon.name))
-			fail_message(user, "You failed imbuing [attacking_weapon]...")
+	if(is_strong_fuel)
+		if(forge_fuel_strong >= 5 MINUTES)
+			fail_message(user, "[src] is full on coal")
 			return
+	if(forge_fuel_weak >= 5 MINUTES)
+		fail_message(user, "[src] is full on wood")
+		return
 
-		attacking_weapon.color = mix_color_from_reagents(attacking_weapon.reagents.reagent_list)
-		to_chat(user, span_notice("You finish imbuing [attacking_weapon]..."))
-		user.mind.adjust_experience(/datum/skill/smithing, 60) //successfully imbuing will grant great experience!
-		playsound(src, 'sound/magic/demon_consume.ogg', 50, TRUE)
-		in_use = FALSE
-		return TRUE
+	balloon_alert_to_viewers("refueling...")
 
-	if(attacking_item.GetComponent(/datum/component/reagent_clothing))
-		var/obj/item/attacking_clothing = attacking_item
+	var/obj/item/stack/sheet/stack_sheet = refueling_stack
+	if(!do_after(user, 3 SECONDS, target = src) || !stack_sheet.use(1))
+		fail_message(user, "stopped fueling")
+		return
 
-		if(in_use) //only insert one at a time
-			to_chat(user, span_warning("You cannot do multiple things at the same time!"))
-			return
-		in_use = TRUE
+	if(is_strong_fuel)
+		forge_fuel_strong += 5 MINUTES
+	else
+		forge_fuel_weak += 5 MINUTES
+	in_use = FALSE
+	balloon_alert(user, "fueled [src]")
+	user.mind.adjust_experience(/datum/skill/smithing, 5) // You gain small amounts of experience from useful fueling
 
-		if(!reagent_forging)
-			fail_message(user, "You must enchant [src] to allow reagent imbuing!")
-			return
+	if(prob(45) && !is_strong_fuel)
```

Comment:
Likewise, the system for making coal should probably be up to more than random chance and a minute of waiting.